src/plugins/dashboard/public/application/utils/use/use_dashboard_container.tsx

```diff
@@ -0,0 +1,459 @@
+  openAddPanelFlyout,
+} from '../../../embeddable_plugin';
+import {
+  convertPanelStateToSavedDashboardPanel,
+  convertSavedDashboardPanelToPanelState,
+} from '../../lib/embeddable_saved_object_converters';
+import { DashboardEmptyScreen, DashboardEmptyScreenProps } from '../../dashboard_empty_screen';
+import {
+  DashboardAppState,
+  DashboardAppStateContainer,
+  DashboardServices,
+  SavedDashboardPanel,
+} from '../../../types';
+import { migrateLegacyQuery } from '../../lib/migrate_legacy_query';
+import { getSavedObjectFinder } from '../../../../../saved_objects/public';
+import { DashboardConstants } from '../../../dashboard_constants';
+import { SavedObjectDashboard } from '../../../saved_dashboards';
+import { Dashboard } from '../../../dashboard';
+import { updateStateUrl } from '../create_dashboard_app_state';
+
+export const useDashboardContainer = (
+  services: DashboardServices,
+  dashboard?: Dashboard,
+  savedDashboardInstance?: SavedObjectDashboard,
+  appState?: DashboardAppStateContainer
+) => {
+  const [dashboardContainer, setDashboardContainer] = useState<DashboardContainer>();
+  const [indexPatterns, setIndexPatterns] = useState<IndexPattern[]>([]);
+
+  useEffect(() => {
+    const getDashboardContainer = async () => {
+      try {
+        if (savedDashboardInstance && appState && dashboard) {
+          const dashboardContainerEmbeddable = await createDashboardEmbeddable(
+            savedDashboardInstance,
+            services,
+            appState,
+            dashboard,
+            setIndexPatterns
+          );
+
+          setDashboardContainer(dashboardContainerEmbeddable);
+        }
+      } catch (error) {
+        services.toastNotifications.addWarning({
+          title: i18n.translate('dashboard.createDashboard.failedToLoadErrorMessage', {
+            defaultMessage: 'Failed to load the dashboard',
+          }),
+        });
+        services.history.replace(DashboardConstants.LANDING_PAGE_PATH);
+      }
+    };
+
+    getDashboardContainer();
+  }, [savedDashboardInstance, appState, services, dashboard]);
+
+  useEffect(() => {
+    const incomingEmbeddable = services.embeddable
+      .getStateTransfer(services.scopedHistory)
+      .getIncomingEmbeddablePackage();
+
+    if (
+      incomingEmbeddable &&
+      !dashboardContainer?.getInput().panels[incomingEmbeddable.embeddableId!]
+    ) {
+      dashboardContainer?.addNewEmbeddable<EmbeddableInput>(
+        incomingEmbeddable.type,
+        incomingEmbeddable.input
+      );
+    }
+  }, [dashboardContainer, services]);
+
+  return { dashboardContainer, indexPatterns };
+};
+
+const createDashboardEmbeddable = (
+  savedDash: any,
+  dashboardServices: DashboardServices,
+  appState: DashboardAppStateContainer,
+  dashboard: Dashboard,
+  setIndexPatterns: React.Dispatch<React.SetStateAction<IndexPattern[]>>
+) => {
+  let dashboardContainer: DashboardContainer;
+  let inputSubscription: Subscription | undefined;
```

Comment:
I see a type complains `'inputSubscription' is declared but its value is never read.ts(6133) ` and think it is because this `inputSubscription` is assigned a value, but its value is never read or used. Same for `outputSubscription`.